### PR TITLE
Switch from modbus to OPC-UA for MPS communication

### LINKED ISSUE #14: Refbox OPC-UA communication
Replace the communication to the MPS machines with a OPC-UA communication based on [open62541](https://github.com/open62541/open62541).

For testing, use the mockup developed in robocup-logistics/rcll-mps-stations#1.

### PR DESCRIPTION
**Pull request by morxa**
Switch to OPC-UA and a new communication protocol for the communication with the MPS machines. In particular:

### library `mps_comm`
Implement OPC-UA communication in the new library `mps_comm`. The library provides commands such as `conveyor_move` and `get_base`, which are provided to CLIPS by function bindings. All commands are processed in a command queue by a separate worker thread. The worker thread waits 40ms after sending a command to make sure that the MPS does not get two commands within 20ms (in which case the first command would be ignored). Commands are run non-blocking, i.e., we do not wait for the MPS to complete the command. Instead, also register callbacks for the `BUSY` and `READY` flags, which can then be processed by the refbox (in CLIPS).

The new library supports three connection modes for an MPS:
* `MOCKUP`: Do not actually connect to any endpoint, but simply mock commands by blocking for a fixed amount of time.
* `SIMULATION`: Connect to a [simulated MPS](https://github.com/robocup-logistics/rcll-mps-stations#mps_sim), which allows setting the registers from a web interface. I've never used this personally, but it was used by @maestrini to develop the OPC-UA communication.
* `PLC`: Connect to a PLC on an actual machine.

The modes can be configured, see the provided config for an example.

For the actual connection to the MPS, use freeopcua. Note that currently there is no Ubuntu package for that library. The circleci config is adapted accordingly to build with freeopcua on Fedora, and to not fail due to the missing library on Ubuntu.

### Refbox changes in CLIPS
Implement the new communication logic in CLIPS, which requires a number of changes to the production code. We now instruct a machine for each step, e.g., first move the conveyor belt to the middle, then retrieve a cap, then move the workpiece to the output. The pattern is similar for each MPS, e.g., for a cap station:

1. When receiving a prepare, start the conveyor belt.
2. Wait for the machine to become `BUSY` and then not `BUSY`. Set the machine to `PROCESSING`.
3. Instruct the MPS to mount/receive a cap.
4. Wait for the machine to be not `BUSY`. Set the state to `PROCESSED`.
5. Move the workpiece to the output.
6. Wait for the machine to be `READY`. Set the machine to `READY-AT-OUTPUT`.
7. Wait for `READY` to be `FALSE`. Set the machine back to `IDLE`.

Along the way, get rid of the slots `proc-state`, `mps-state`, and `mps-state-deferred`. The MPS no longer has a separate state, we track everything on our side now. The slot `proc-state` became unnecessary due to the other changes. Also get rid of most of the simulation code, as this is no longer necessary and replaced by the mockup functionality of `mps_comm`.

This resolves #14 and supersedes #21.

Co-authored by @maestrini.

### CODE REVIEW

File: src/libs/mps_comm/machine.cpp [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+          }
+          return;
+        }
+				OpcUtils::MPSRegister registerOffset;
+				if (command < Station::STATION_BASE)
+					registerOffset = OpcUtils::MPSRegister::ACTION_ID_BASIC;
+				else
+					registerOffset = OpcUtils::MPSRegister::ACTION_ID_IN;
+
+				bool                  statusBit = (bool)(status & Status::STATUS_BUISY);
```
Comment (TarikViehmann) [inline src/libs/mps_comm/machine.cpp]:
```suggestion
				bool                  statusBit = (bool)(status & Status::STATUS_BUSY);
```
That looks like a typo.

File: src/libs/mps_comm/machine.cpp [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+}
+
+void Machine::wait_for_buisy() {
+  struct timespec time_c, time_0, time_d;
+  clock_gettime(CLOCK_MONOTONIC, &time_0);
+    
+  while ((getReturnValue(OpcUtils::MPSRegister::STATUS_BUSY_IN)->ToInt())) {
+    clock_gettime(CLOCK_MONOTONIC, &time_c);
+    timespec_diff( &time_0, &time_c, &time_d);
+    if( time_d.tv_sec * 1000 + time_d.tv_nsec / 1000000 > Timeout::TIMEOUT_BUISY*10) { /* TODO REMOVE *10 from timeout */
```
Comment (TarikViehmann) [inline src/libs/mps_comm/machine.cpp]:
```suggestion
    if( time_d.tv_sec * 1000 + time_d.tv_nsec / 1000000 > Timeout::TIMEOUT_BUSY*10) { /* TODO REMOVE *10 from timeout */
```
Another typo here i guess

File: src/libs/mps_comm/machine.cpp [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+      }
+    }
+  }
+  if (getReturnValue(OpcUtils::MPSRegister::STATUS_ENABLE_IN)->ToInt())
+    return true;
+  else
+    return false;
+}
+
+void Machine::wait_for_buisy() {
```
Comment (TarikViehmann) [inline src/libs/mps_comm/machine.cpp]:
```suggestion
void Machine::wait_for_busy() {
```
Should also be changed in the header, of course.

File: src/libs/mps_comm/machine.cpp (hunk 1 above)
Comment (morxa) [reply src/libs/mps_comm/machine.cpp]: Thanks for pointing those out, I fixed all of them.

**Comment by morxa** [comment]
After @TarikViehmann's comments, I realized there were still active polling implementations for `wait_for_{ready,busy,free}`. I removed those, as they are not used anymore and do not work well (due to active polling).

**Comment by MostafaGomaa** [comment]
So I have cartfully went through the changes in production.clp (in parallel with and after re-reading the old production.clp to have it fresh in memory)

I have to say, not only is the flow is much much clearer now, many bugs and workarounds form that past are also remedied. Thank you for that. Great work indeed decoding that to this!!

Yet in going though it,  I encountered some minor styling and readability issues that I think could be imporved to help get the gest faster.  I will point those out shortly with the lable "suggestion" since its not crutial to the functionlity and could be added later (even though I think it would be better to handle them sooner)
Those mainly are:
- Many of the useful printing is either gone or not added 
- Sometimes the order of the rules seems a bit random which causes to keep jumping up and down to find related rules (I suppose this is due the speed and incrimental nature or robocup)
- Some of the documentation and the rules names could be better

Follows, please find my change requests and inline minor suggestions

**Comment by morxa** [comment]
Thanks for the comments!

> So I have carefully gone through the changes in production.clp (in parallel with and after re-reading the old production.clp to have it fresh in memory)
> 
> I have to say, not only is the flow is much much clearer now, many bugs and workarounds form that past are also remedied. Thank you for that. Great work indeed decoding that to this!!
> 
> Yet in going through it, I encountered some minor styling and readability issues that I think could be improved to help get the gest faster. I will point those out shortly with the labeled "suggestion" or "styling" since its not crucial to the functionality and could be added later (even though I think it would be better to handle them sooner)
> Those mainly are:
> 
>     * Many of the useful printing is either gone or not added

I deliberately removed some output because it was redundant. What exactly are you missing?

> 
>     * Sometimes the order of the rules seems a bit random which causes to keep jumping up and down to find related rules (I suppose this is due to the speed and incremental nature of RoboCup)

Most rules are ordered by mps-type. As some rules apply to multiple types, they may seem out-of-place, but that shouldn't be a big issue. Do you have a specific suggestion for reordering?
> 
>     * Some of the documentation and the names of the rules could be improved

Done.

> 
> 
> Follows, please find my change requests and inline minor suggestions

I also removed a redundant rule and silenced the output by OPC-UA a bit.

File: src/games/rcll/production.clp [hunk 4]
```diff
@@ -255,339 +268,307 @@
 
 ; **** Machine state processing
 
-(defrule prod-proc-state-idle
-  (gamestate (state RUNNING) (phase PRODUCTION) (game-time ?gt))
-  ?m <- (machine (name ?n) (state IDLE) (proc-state ~IDLE))
+(defrule production-green-light-on-idle
+  (gamestate (state RUNNING) (phase PRODUCTION))
+  ?m <- (machine (name ?n) (state IDLE) (actual-lights $?lights&:(neq ?lights (create$ GREEN-ON))))
```
[the remark this answers is not in the document]
Comment (morxa) [reply src/games/rcll/production.clp]: I moved all machine light handling into separate rules and into a separate file.

File: src/games/rcll/production.clp [hunk 5]
```diff
@@ -255,339 +268,307 @@
...
-  "BS must be instructed to dispense base for processing"
-  (declare (salience ?*PRIORITY_HIGH*))
-  (gamestate (state RUNNING) (phase PRODUCTION) (game-time ?gt))
-  ?m <- (machine (name ?n) (mtype DS) (state PROCESSING) (proc-state ~PROCESSING)
-		 (ds-gate ?gate))
-  =>
-  (modify ?m (proc-state PROCESSING) (desired-lights GREEN-ON YELLOW-ON) (ds-gate 0) (ds-last-gate ?gate))
+(defrule production-bs-idle
+	"The base has been picked up"
+	?m <- (machine (name ?n) (mtype BS|RS) (state READY-AT-OUTPUT) (task MOVE-OUT) (mps-ready FALSE))
```
[the remark this answers is not in the document]
Comment (morxa) [reply src/games/rcll/production.clp]: Good catch, I removed the redundant rule!

We currently do not have `WAIT-IDLE` anymore. If I understood correctly, this was just waiting for the machine to report itself as `IDLE` and timeout if that didn't happen. But I'm not entirely sure if that's correct. Anyway, I don't see a purpose in blinking yellow after the workpiece has been removed, do you?

File: src/games/rcll/production.clp [hunk 6]
```diff
@@ -48,15 +48,26 @@
...
-   else (modify ?mf (state ?prev-state) (proc-state DOWN)
-		(mps-state ?mps-state) (mps-state-deferred NONE))
-  )
+	(printout t "Machine " ?name " is up again" crlf)
+	(modify ?mf (state ?prev-state))
+)
+
+(defrule prod-mps-state-ready
+	?m <- (machine (name ?n))
+  ?mps-status <- (mps-status-feedback ?n READY ?ready)
```
[the remark this answers is not in the document]
Comment (morxa) [reply src/games/rcll/production.clp]: Fixed

File: src/games/rcll/production.clp [hunk 7]
```diff
@@ -255,339 +268,307 @@
...
-  ?m <- (machine (name ?n) (mtype RS) (state PROCESSED) (proc-state ~PROCESSED)
-		 (rs-ring-color ?ring-color) (bases-used ?bu))
-  (ring-spec (color ?ring-color) (req-bases ?req-bases))
-  =>
-  (printout t "Machine " ?n " finished processing, moving to output" crlf)
-  (modify ?m (proc-state PROCESSED) (bases-used (+ ?bu ?req-bases)))
-  (mps-deliver (str-cat ?n))
+(defrule production-cs-cap-main-op
+	?m <- (machine (name ?n) (mtype CS) (state PREPARED) (task MOVE-MID) (mps-busy FALSE)
+	               (cs-operation ?cs-op))
```
[the remark this answers is not in the document]
Comment (morxa) [reply src/games/rcll/production.clp]: Well, for the CS there are two operations, that's why it's called main-op.

File: src/games/rcll/production.clp [hunk 8]
```diff
@@ -255,339 +268,307 @@
...
+	(modify ?m (state IDLE) (prev-state BROKEN) (bases-used ?ba) (cs-retrieved FALSE) (broken-since 0.0))
 )
 
-(defrule prod-prepared-but-no-input
+(defrule production-prepared-but-no-input
   (gamestate (state RUNNING) (phase PRODUCTION) (game-time ?gt))
-  ?m <- (machine (name ?n) (mtype ~BS) (state PREPARED|PROCESSING|PROCESSED)
-        (wait-for-product-since ?ws&:(timeout-sec ?gt ?ws ?*PREPARE-WAIT-TILL-RESET*)))
+  ?m <- (machine (name ?n) (mtype ?type&~BS) (state ?state
+	               &:(or (and (member$ ?type (create$ DS BS)) (eq ?state PROCESSING)) (eq ?state PREPARED)))
```
[the remark this answers is not in the document]
Comment (morxa) [reply src/games/rcll/production.clp]: True, allowing the BS was unnecessary. But the reasoning is different: In contrast to the other machines, the DS directly switches to `PROCESSING` while waiting for an input. Thus, if there is no workpiece fed into the machine, the machine should break, even if it already `PROCESSING`.

File: src/games/rcll/production.clp [hunk 9]
```diff
@@ -255,339 +268,307 @@
...
-  (gamestate (state RUNNING) (phase PRODUCTION) (game-time ?gt))
-  ?m <- (machine (name ?n) (state WAIT-IDLE)
-		 (retrieved-at ?r&:(timeout-sec ?gt ?r ?*RETRIEVE-WAIT-IDLE-TIME*)))
-  =>
-  (printout t "retrieval timeout, going to IDLE" crlf)
-  (modify ?m (state IDLE) (proc-state WAIT-IDLE))
+(defrule production-mps-product-retrieved
+	?m <- (machine (name ?n) (state READY-AT-OUTPUT) (mps-ready FALSE))
+	=>
+	(modify ?m (state IDLE) (desired-lights GREEN-ON))
```
[the remark this answers is not in the document]
Comment (morxa) [reply src/games/rcll/production.clp]: See above for why we don't have `WAIT-IDLE` anymore.

File: src/games/rcll/production.clp [hunk 10]
```diff
@@ -255,339 +268,307 @@
...
-  ?m <- (machine (name ?n) (state READY-AT-OUTPUT) (mps-state RETRIEVED)
-		 (proc-time ?pt) (proc-start ?pstart&:(timeout-sec ?gt ?pstart ?pt)))
-  =>
-  (modify ?m (state WAIT-IDLE) (retrieved-at ?gt) (desired-lights YELLOW-BLINK))
+  ?m <- (machine (name ?n) (state PROCESSING|PROCESSED)
+        (wait-for-product-since ?ws
+	        &:(timeout-sec ?gt ?ws (+ ?*PREPARE-WAIT-TILL-RESET* ?*PROCESSING-WAIT-TILL-RESET*))))
+	=>
+	(printout error "Machine " ?n " timed out while processing" crlf)
+	(modify ?m (state IDLE) (task nil))
```
[the remark this answers is not in the document]
Comment (morxa) [reply src/games/rcll/production.clp]: I initially thought that if a machine breaks while it is processing, it's always a machine failure. But that's not true, so I changed it so the machine will go to `BROKEN` instead. If it actually is a machine failure, the team will probably file a complaint anyway, so there is no reason not to set the machine to `BROKEN`.

I also used `proc-start` for the timeout, thanks for the suggestion!

**Comment by morxa** [comment]
Thanks for the thorough reviews! I tried to tackle all concerns, please review again.

File: src/games/rcll/production.clp (hunk 5 above)
[the remark this answers is not in the document]
Comment (MostafaGomaa) [reply src/games/rcll/production.clp]: I agree that "WAIT-IDLE" is not absolutely necessary it was always a transient state that times out after READY_AT_OUTPUT (it was never related to mps states, the mps-state had no notion of an IDLE state). 
On the other hand,  It was used to indicate to the spectator that the production at that machine has been completed correctly and distinguishes the upcoming IDLE state (that preserves the preparation status done at a machine). This distinguishes to the spectator the reason of being IDLE is a completed production (as opposed to any other possible machine reset for instance which would not preserve the prepared status of the machine and will completely reset it )

So in my opinion, as the use case for it never really changed by the filed changes, I think keeping the wait-idle does if anything is a useful visual aid.
that said, I would not be completely opposed to getting rid of it if you want, I just really think it's useful and does not harm

**Comment by MostafaGomaa** [comment]
Thank you for the quick and rather precise edits. I will take another look asap

**Comment by MostafaGomaa** [comment]
Other than the pending comments about the WAIT-IDLE status I have no further comments about this PR.
I will gladly approve this PR once that is resolved.
Thank you very much for your effort making this happen and neatly realizing the changes arising during reviewing of the PR :)

File: src/games/rcll/production.clp (hunk 5 above)
[the remark this answers is not in the document]
Comment (morxa) [reply src/games/rcll/production.clp]: Makes sense, I re-introduced the state `WAIT-IDLE`.

**Comment by morxa** [comment]
> Other than the pending comments about the WAIT-IDLE status I have no further comments about this PR.
> I will gladly approve this PR once that is resolved.
> Thank you very much for your effort making this happen and neatly realizing the changes arising during reviewing of the PR :)

I re-added `WAIT-IDLE`. Please have another look!

**Pull request merged**